sound/soc/soc-pcm.c

```diff
@@ -2918,30 +2919,41 @@ int soc_new_pcm(struct snd_soc_pcm_runtime *rtd, int num)
 	}
 
 	if (rtd->dai_link->dynamic || rtd->dai_link->no_pcm) {
-		for_each_rtd_cpu_dais(rtd, i, cpu_dai) {
+		if (rtd->dai_link->dpcm_playback) {
+			invalid_dais = 0;
+			stream = SNDRV_PCM_STREAM_PLAYBACK;
+
+			for_each_rtd_cpu_dais(rtd, i, cpu_dai)
+				if (!snd_soc_dai_stream_valid(cpu_dai, stream))
+					invalid_dais++;
 
-			if (rtd->dai_link->dpcm_playback) {
-				stream = SNDRV_PCM_STREAM_PLAYBACK;
-				if (snd_soc_dai_stream_valid(cpu_dai, stream)) {
-					playback = 1;
-				} else if (playback != 0) {
```

Comment:
As far as I understand the only difference with the current version is, that if currently, say, in this case of `rtd->dai_link->dpcm_playback != 0` several of the first CPU DAIs are "invalid" for playback, the checks will not identify them because of this `else if`. So the current version is too permissive, it fails to identify some invalid configurations, and your patch fixes that? If this my understanding is correct, then the simplest fix would be just
```
-	} else if (playback != 0) {
+	} else {
```
and similar below for capture? This would also avoid looping twice over CPU DAIs? A slightly bigger diff but also "prettifying" the code a bit would be
```
-				if (snd_soc_dai_stream_valid(cpu_dai, stream)) {
-					playback = 1;
-				} else if (playback != 0) {
-					dev_err(rtd->card->dev,
-						"CPU DAI %s does not support playback\n",
-						cpu_dai->name);
-					return -EINVAL;
-				}
+				if (!snd_soc_dai_stream_valid(cpu_dai, stream)) {
+					dev_err(rtd->card->dev,
+						"CPU DAI %s does not support playback\n",
+						cpu_dai->name);
+					return -EINVAL;
+				}
+				playback = 1;
```

Reply:
yes, much better, thanks @lyakh for the suggestion.